### PR TITLE
PWX-30276: Do not delete pods running on storage down nodes.

### PR DESCRIPTION
**Pull request by adityadani**
**What type of PR is this?**
> Uncomment only one and also add the corresponding label in the PR:
>bug

**What this PR does / why we need it**:
- Introduce a new state in Stork for storage down nodes.
- Remap portworx StorageDown state to Stork's StorageDown state
- For degraded node state:
  - filter out the nodes from extender requests.
  - monitor should delete the pods running on them
- For storage down state:
  - do not filter out such nodes.
  - give a low score to nodes which are storage down.
  - monitor should not delete pods running on them

**Does this PR change a user-facing CRD or CLI?**:
No

**Is a release note needed?**:

```release-note
Issue: Stork would delete pods running on Degraded or Portworx StorageDown nodes.
User Impact: Applications would face disruption even though drivers like Portworx supported applications running on StorageDown nodes.
Resolution: Stork will not delete pods running on StorageDown nodes.

```

**Does this change need to be cherry-picked to a release branch?**:
Yes - 23.4

**Testing Notes**
- Added a UT to ensure pods are not deleted on StorageDown nodes.
- Modified existing UTs to use StorageDown state instead of NodeDegraded state. 
- Added a UT in extender to not schedule pods on Degraded nodes.
- Currently working on adding integration tests.

### CODE REVIEW

File: pkg/monitor/monitor.go
Comment (pp511) [inline pkg/monitor/monitor.go:213]: During exponential backoff we check for node condition again here 
https://github.com/libopenstorage/stork/blob/master/pkg/monitor/monitor.go#:~:text=if%20n.Status%20!%3D%20volume.NodeOnline%20%7B
This condition should be updated too.

File: pkg/extender/extender_test.go
Comment (pp511) [inline pkg/extender/extender_test.go:371]: nit.. StorageDown->NodeDegraded

File: pkg/extender/extender_test.go
Comment (pp511) [inline pkg/extender/extender_test.go:408]: I think we will need to call CreateCluster(6, nodes) above to simulate a 6 node cluster with 1 node in node degraded state.

File: pkg/extender/extender_test.go
Comment (pp511) [inline pkg/extender/extender_test.go:409]: nit.. StorageDown->NodeDegraded

File: pkg/monitor/monitor_test.go
Comment (pp511) [inline pkg/monitor/monitor_test.go:319]: Nit..Degraded->StorageDown

**Review by pp511 — CHANGES_REQUESTED** [review]
Overall looks good.
Some nitpicks.
One comment in monitor.go needs to be addressed before we merge.

File: pkg/monitor/monitor_test.go
Comment (pp511) [reply pkg/monitor/monitor_test.go:319]: You missed updating this.

**Review by pp511 — APPROVED** [review]
lgtm. just one nit that you missed addressing. ok to merge after CBTs pass.

**Comment by adityadani** [comment]
Thanks @pp511 ! Will fix the missed comment. Currently trying to get a successful integration test run with the newly added tests. Will merge it once I get a success.

File: test/integration_test/extender_test.go
Comment (pp511) [inline test/integration_test/extender_test.go:263]: nit..Error exiting pool maintenance

File: test/integration_test/extender_test.go
Comment (pp511) [inline test/integration_test/extender_test.go:242]: Nit..Error stopping driver -> Error entering pool maintenance

**Comment by pp511** [comment]
Integration tests look good. Just two nits that need to be addressed

**Pull request merged**
